contracts/contracts/SiweParser.sol

```diff
@@ -0,0 +1,252 @@
+            if (str[j] == 0x0a) {
+                return (_substr(str, i, j), j + 1);
+            }
+        }
+        return (_substr(str, i, j), j);
+    }
+
+    /**
+     * @notice Parse bullets, one per line in str starting at i.
+     * @return Array of parsed values and a new i.
+     */
+    function _parseArray(bytes calldata str, uint256 i)
+        internal
+        pure
+        returns (bytes[] memory, uint256)
+    {
+        // First count the number of resources.
+        uint256 j = i;
+        uint256 count = 0;
+        for (; j < str.length - 1; j++) {
+            if (str[j] == "-" && str[j + 1] == " ") {
+                j += 2;
+                count++;
+            } else {
+                break;
+            }
+            while (j < str.length && str[j] != 0x0a) {
+                j++;
+            }
+        }
+
+        // Then build an array.
+        bytes[] memory values = new bytes[](count);
+        j = i;
+        for (uint256 c = 0; j < str.length - 1 && c != count; j++) {
+            if (str[j] == "-" && str[j + 1] == " ") {
+                i = j + 2;
+            }
+            while (j < str.length && str[j] != 0x0a) {
+                j++;
+            }
+            values[c] = _substr(str, i, j);
+            c++;
+            if (j == str.length) {
+                j--; // Subtract 1 because of the outer loop.
+            }
+        }
+        return (values, j);
+    }
+
+    /**
+     * @notice Parse SIWE message.
+     * @return ParsedSiweMessage struct with populated fields from the message.
+     */
+    function parseSiweMsg(bytes calldata siweMsg)
+        internal
+        pure
+        returns (ParsedSiweMessage memory)
+    {
+        ParsedSiweMessage memory p;
+        uint256 i = 0;
+
+        // dApp Domain.
+        for (; i < siweMsg.length; i++) {
+            if (siweMsg[i] == " ") {
+                p.schemeDomain = _substr(siweMsg, 0, i);
+                break;
+            }
+        }
+
+        i += 50; // " wants you to sign in with your Ethereum account:\n"
+
+        // Address.
+        // TODO: Verify the mixed-case checksum.
+        p.addr = _hexStringToAddress(_substr(siweMsg, i += 2, i += 40));
+        i += 2; // End of address new line + New line.
+
+        // (Optional) statement.
+        if (i < siweMsg.length && siweMsg[i] != "\n") {
+            for (uint256 j = i; j < siweMsg.length; j++) {
+                if (siweMsg[j] == 0x0a) {
+                    p.statement = _substr(siweMsg, i, j);
+                    i = j + 1; // End of statement new line.
+                    break;
+                }
+            }
+        }
+
+        i++; // New line.
+
+        (p.uri, i) = _parseField(siweMsg, "URI", i);
+        (p.version, i) = _parseField(siweMsg, "Version", i);
+        bytes memory chainId;
+        (chainId, i) = _parseField(siweMsg, "Chain ID", i);
+        p.chainId = _parseUint(chainId);
+        (p.nonce, i) = _parseField(siweMsg, "Nonce", i);
```

Comment:
> nonce REQUIRED. A random string typically chosen by the relying party and used to prevent replay attacks, at least 8 alphanumeric characters.

Probably too pedantic?